hail/src/main/scala/is/hail/expr/ir/Emit.scala

```diff
@@ -1996,21 +2046,139 @@ private class Emit(
+        val rP = rChild.pType.asInstanceOf[PNDArray]
+
+        val lElemType = lP.elementType
+        val rElemType = rP.elementType
+
+        val lVti = typeToTypeInfo(lElemType.virtualType)
+        val rVti = typeToTypeInfo(rElemType.virtualType)
+
+        val lElemRef = coerce[Any](mb.newField(lName)(lVti))
+        val rElemRef = coerce[Any](mb.newField(rName)(rVti))
+
+        val bodyEnv = env.bind(name=lName, v=(lVti, false, lElemRef.load()))
+                         .bind(name=rName, v=(rVti, false, rElemRef.load()))
+        val bodyt = this.emit(body, bodyEnv, er, None)
+
+        val leftChildEmitter = deforest(lChild)
+        val rightChildEmitter = deforest(rChild)
+
+        val (unifiedShapeSetup, unifiedShape, unifiedShapePType) = NDArrayEmitter.unifyShapes(mb, leftChildEmitter.outputShape, leftChildEmitter.outputShapePType,
+          rightChildEmitter.outputShape, rightChildEmitter.outputShapePType, region)
+
+        val setup = Code(leftChildEmitter.setup, rightChildEmitter.setup, unifiedShapeSetup)
+
+        new NDArrayEmitter(mb, unifiedShapePType.size, unifiedShape,
+          lP.shape.pType, body.pType, setup) {
+          override def outputElement(idxVars: Array[Code[Long]]): Code[_] = {
+
+            val lIdxVars2 = NDArrayEmitter.zeroBroadcastedDims2(mb, idxVars, nDims, leftChildEmitter.outputShape, leftChildEmitter.outputShapePType)
+            val rIdxVars2 = NDArrayEmitter.zeroBroadcastedDims2(mb, idxVars, nDims, rightChildEmitter.outputShape, rightChildEmitter.outputShapePType)
+
+            Code(
+              lElemRef := leftChildEmitter.outputElement(lIdxVars2),
+              rElemRef := rightChildEmitter.outputElement(rIdxVars2),
+              bodyt.setup,
+              bodyt.m.orEmpty(Code._fatal("NDArray map body cannot be missing")),
+              bodyt.v
+            )
+          }
+        }
+
+      case x@NDArrayReindex(child, indexExpr) =>
+        val childEmitter = deforest(child)
+        val childPType = child.pType.asInstanceOf[PNDArray]
+
+        val outputPType = x.pType
+        val outputShapePType = outputPType.representation.fieldType("shape").asInstanceOf[PTuple]
+
+        val shapeSrvb = new StagedRegionValueBuilder(mb, outputShapePType)
+
+        val childShapeTuple = new CodePTuple(childPType.shape.pType, region, childEmitter.outputShape)
+
+        val reindexShape = indexExpr.map {childIndex =>
+          if (childIndex < childPType.nDims) {
+            Code(
+              shapeSrvb.addLong(childShapeTuple(childIndex)),
+              shapeSrvb.advance()
+            )
+          }
+          else {
+            Code(
+              shapeSrvb.addLong(1L),
+              shapeSrvb.advance()
+            )
+          }
+        }
+
+        val setup = Code(childEmitter.setup, shapeSrvb.start(), reindexShape)
+
+        new NDArrayEmitter(mb, indexExpr.length, shapeSrvb.end(), outputShapePType, outputPType.elementType, setup) {
```

Comment:
oof, sorry again, we can't be doing this.

We're not "deforesting" if every IR node could potentially allocate a shape tuple!

NDArrayEmitter probably should take an `SRVB => Code[Unit]` which will add the shape to the ndarray we're building.

Reply:
`makeDefaultStrides` needs to be changed to use an SRVB as well, it shouldn't be manipulating the region directly like that.

Reply:
OK, so I think each NDArrayEmitter should just define for itself what its shape *and* strides are. Trying to do it at the end is gonna be a mess because we want to look at the shape of an NDArray that we're currently building using an SRVB.

Reply:
Can we make them `Array[Code[Long]]`? I think that's actually the right thing.

Reply:
You want the shape and the strides to be passed along as `Array[Code[Long]]`, and then I only realize one shape tuple at the end? That should be fine.

Reply:
Just to be clear, the elements of this array are Code's representing the longs that will make up the shape?